### PR TITLE
#35 login

### PR DESCRIPTION
**Pull request by AdamBlazsek**
Ready for review

### CODE REVIEW

File: Phaethon/Core/Model/Login.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+using System.Runtime.Serialization;
+
+namespace Core.Model
+{
+    [Serializable]
+    [DataContract]
+    public class Login
+    {
```
Comment (RaidenRabit) [inline Phaethon/Core/Model/Login.cs]: https://github.com/RaidenRabit/Phaethon/blob/da6d6b8edb567c6e375539f3d112132bbce69993/Phaethon/Core/Model/Login.cs#L13-L23
If Username is unique, why do we need int id? Cant we just use username as id?

File: Phaethon/InternalApi/Controllers/LoginController.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            if(loginID !=0)
+                return Request.CreateResponse(HttpStatusCode.OK, _loginManagement.Login(login.Username, login.Password));
+            else
+            {
+                return Request.CreateResponse(HttpStatusCode.BadRequest, "Incorrect login credentials");
+            }
+        }
+
+    }
+}
```
Comment (RaidenRabit) [inline Phaethon/InternalApi/Controllers/LoginController.cs]: All of these methods will have to be refactored, to actually use the login.

For example: you dont want random people to make a call to any of those endpoints, with correct queries, and get data, unless they are logged in.

File: Phaethon/InternalApi/DataManagement/IDataManagement/ILoginManagement.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Core.Model;
+
+namespace InternalApi.DataManagement.IDataManagement
+{
+    interface ILoginManagement
```
Comment (RaidenRabit) [inline Phaethon/InternalApi/DataManagement/IDataManagement/ILoginManagement.cs]: Please rename it to ILoginDM

File: Phaethon/InternalApi/DataManagement/LoginManagement.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+using System.Text;
+using System.Web;
+using System.Web.Security;
+using Core.Model;
+using InternalApi.DataAccess;
+using InternalApi.DataManagement.IDataManagement;
+
+namespace InternalApi.DataManagement
+{
+    public class LoginManagement : ILoginManagement
```
Comment (RaidenRabit) [inline Phaethon/InternalApi/DataManagement/LoginManagement.cs]: Please rename it to LoginDM

File: Phaethon/InternalApi/DataManagement/LoginManagement.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using System.Collections.Generic;
```
Comment (RaidenRabit) [inline Phaethon/InternalApi/DataManagement/LoginManagement.cs]: Please keep the references to a minimum

File: Phaethon/InternalApi/DataManagement/LoginManagement.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                using (var db = new DatabaseContext())
+                {
+                    login.Salt = GenerateSalt();
+                    login.Password =
+                        Convert.ToBase64String(ComputeHMAC_SHA256(Encoding.UTF8.GetBytes(login.Password), login.Salt));
+                    _LoginDa.CreateOrUpdate(db, login);
+                    return true;
+                }
+            }
+            catch (Exception e)
```
Comment (RaidenRabit) [inline Phaethon/InternalApi/DataManagement/LoginManagement.cs]: You're doing nothing with e, so you dont need it.

File: Phaethon/InternalApi/DataManagement/LoginManagement.cs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+                    }
+                    else
+                    { 
+                        return 0;
+                    }
+                }
+                 return 0;
+                
+            }
```
Comment (RaidenRabit) [inline Phaethon/InternalApi/DataManagement/LoginManagement.cs]: https://github.com/RaidenRabit/Phaethon/blob/da6d6b8edb567c6e375539f3d112132bbce69993/Phaethon/InternalApi/DataManagement/LoginManagement.cs#L74-L86

You don't need those. Delete everything but line 84.

File: Phaethon/InternalApi/Global.asax.cs [hunk 8]
```diff
@@ -12,12 +14,14 @@ public class WebApiApplication : HttpApplication
     {
         protected void Application_Start()
         {
+
```
Comment (RaidenRabit) [inline Phaethon/InternalApi/Global.asax.cs]: random new lines?

File: Phaethon/InternalApi/Controllers/LoginController.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
+﻿using System;
+using System.Collections.Generic;
```
Comment (RaidenRabit) [inline Phaethon/InternalApi/Controllers/LoginController.cs]: Remove unecessary references

File: Phaethon/InternalApi/DataAccess/LoginDa.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
+﻿using System;
+using System.Security.Cryptography;
```
Comment (RaidenRabit) [inline Phaethon/InternalApi/DataAccess/LoginDa.cs]: Remove unnecessary references

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 11]
```diff
@@ -0,0 +1,202 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: Remove unnecessary references

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+using NUnit.Framework;
+using System.Security.Cryptography;
+using InternalApi.DataManagement;
+
+namespace Tests.IntegrationTests
+{
+
+   public class LoginTest : InternalTestFakeServerBase
+    {
+        private Login userModel;
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: these varaible names should start with _
its just naming convention thing.

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+        private Login userModel2;
+        private int count1;
+        private int count2;
+        private string name;
+        private string name2;
+        private string OGPass;
+        private LoginDa loginDa;
+
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: https://github.com/RaidenRabit/Phaethon/blob/da6d6b8edb567c6e375539f3d112132bbce69993/Phaethon/Tests/IntegrationTests/LoginTest.cs#L26-L30

These should be local variables, as they are used only in 1 test.

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+   public class LoginTest : InternalTestFakeServerBase
+    {
+        private Login userModel;
+        private Login userModel2;
+        private int count1;
+        private int count2;
+        private string name;
+        private string name2;
+        private string OGPass;
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: Bad naming, and why do you need it?

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+        private LoginDa loginDa;
+
+        private void InitializeData()
+        {
+            userModel = new Login { Username = "Subject1", Password = "12355557" };
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: Make your tests independent.

Right now, if one of your tests fails for some reason, Cleanup() will not be reached, and because Username is unique, all subsequent tests will fail.

Perhaps generating new random usernames, for each test would be a good idea.

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+        #region Post
+
+        [Test]
+        public async Task Login_CorrectLoginInfo_OK()
+        {
+            //Setup
+            InitializeData();
+            userModel2.Password = OGPass;
+
+            var response = await _client.PostAsJsonAsync("Login/Login", userModel2);
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: missing the //Act

File: Phaethon/Tests/IntegrationTests/LoginTest.cs (hunk 15 above)
Comment (RaidenRabit) [reply Phaethon/Tests/IntegrationTests/LoginTest.cs]: Or you could wrap your tests in try{} statements and put the Cleanup() in final{}

Or put attibute [TearDown] on your Cleanup() method, so itll be called everytime a test ends, no matter if it fails, encounters error or passes

Or find another way, you're free to do whatever you want.

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+            loginDa = new LoginDa { };
+            
+            using (var db = new DatabaseContext())
+            {
+                Login deleted = db.Login.Attach(userModel2);
+                loginDa.Delete(db, deleted);
+            }
+        }
+
+        #region Post
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: Random region name?

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+                InitializeData();
+           
+            //Act
+           
+            var response = await _client.PostAsJsonAsync("Login/Delete", userModel2.ID.ToString());
+
+            //Assert
+            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
+        }
+        #endregion
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: You're missing a "wrong info" test

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+
+        #region Add
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: random region name?

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+            using (var db = new DatabaseContext())
+            {
+                count2 = db.Login.Count();               
+            }
+            var response = await _client.PostAsJsonAsync("Login/CreateOrUpdate", userModel);
+
+            //Assert
+            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
+            Assert.AreNotEqual(count1, count2);
+            loginDa = new LoginDa { };
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: Normal people do it like: _loginDa = new LoginDa(); not { };

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 21]
```diff
@@ -0,0 +1,202 @@
...
+            using (var db = new DatabaseContext())
+            {
+                Login deleted = db.Login.Attach(userModel);
+                loginDa.Delete(db, deleted);
+            }
+            Cleanup();
+        }
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: Excuse me, but wtf is this test doing?
https://github.com/RaidenRabit/Phaethon/blob/da6d6b8edb567c6e375539f3d112132bbce69993/Phaethon/Tests/IntegrationTests/LoginTest.cs#L132-L158

Comment (RaidenRabit) [reply Phaethon/Tests/IntegrationTests/LoginTest.cs]: image

Well, let me tell you something:
- of course the 2 counts are different, you just inserted 2 users in the db, by directly calling it. Which is fundamentally wrong, as this test is supposed to test if the API is able to insert a user, not if the db can insert user.
- Since Subject1 was inserted with DB query, and Username is supposed to unique, inserting Subject1, using the API should not be possible, since its already there. Meaning theres some API issue as well. (this case is particularly bad, because someone might just know the username of an account, and just create a new user, with the same name and simply replacing the previous one.

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 22]
```diff
@@ -0,0 +1,202 @@
...
+            name = userModel2.Username;
+            userModel2.Username = "Maybe";
+            name2 = userModel2.Username;
+            //Act
+            var response = await _client.PostAsJsonAsync("Login/CreateOrUpdate", userModel2);
+            //Assert
+            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
+            Assert.AreNotEqual(name, name2);
+            Cleanup();
+        }
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: You should add a test for what happens if someone tries to edit their username, but that username already exists in the database.

Normaly, the edit should fail and user should receive a BadRequest with friendly message on why that is not allowed.

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 23]
```diff
@@ -0,0 +1,202 @@
...
+
+        private byte[] ComputeHMAC_SHA256(byte[] data, byte[] salt)
+        {
+            using (var hmac = new HMACSHA256(salt))
+            {
+                return hmac.ComputeHash(data);
+            }
+        }
+        #endregion
+    }
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: https://github.com/RaidenRabit/Phaethon/blob/da6d6b8edb567c6e375539f3d112132bbce69993/Phaethon/Tests/IntegrationTests/LoginTest.cs#L181-L199
Code duplication.

File: Phaethon/Tests/Tests.csproj [hunk 24]
```diff
@@ -264,7 +265,6 @@
   </ItemGroup>
   <ItemGroup>
     <Folder Include="AcceptanceTests\" />
-    <Folder Include="UnitTests\" />
```
Comment (RaidenRabit) [inline Phaethon/Tests/Tests.csproj]: image
Why did you remove the Acceptance and Unit Tests folders?

**Comment by RaidenRabit** [comment]
UI review:
        Missing UI elements:
- Login button
- Logout button
- User Page button

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 25]
```diff
@@ -1,78 +1,78 @@
...
     {
-        private Login userModel;
-        private Login userModel2;
-        private int count1;
-        private int count2;
-        private string name;
-        private string name2;
-        private string OGPass;
-        private LoginDa loginDa;
+        private Login _userModel;
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: this variable is never used?

Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: https://github.com/RaidenRabit/Phaethon/blob/99c8a620e6cbbbe3d1c4f49d3a255f08cea6d221/Phaethon/Tests/IntegrationTests/LoginTest.cs#L18-L21
These variables should be local variables, not class wide variables, because they are used only in 1-2 tests, each.

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 26]
```diff
@@ -1,78 +1,78 @@
...
+        private Login _userModel2;
+        private string _name;
+        private string _name2;
+        private string _originalPass;
+        private LoginDa _loginDa;
 
         private void InitializeData()
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: Add `[SetUp]` attribute to this method (just like you did with TearDown, this way you wont need to call this method in everry test 😸

Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: https://github.com/RaidenRabit/Phaethon/blob/99c8a620e6cbbbe3d1c4f49d3a255f08cea6d221/Phaethon/Tests/IntegrationTests/LoginTest.cs#L144-L156

This test doesn't really make sense, because:
line148: _name = "Subject2";
line149: _userModel2.Username = "Maybe";
line150: _name2 = "Maybe";
line156: Assert.AreNotEqual("Subject2", "Maybe");

What you're doing in those lines, is essencially just comparing 2 different strings, without actually taking the username that was changed, from the DB.

File: Phaethon/Tests/IntegrationTests/LoginTest.cs (hunk 22 above)
Comment (RaidenRabit) [reply Phaethon/Tests/IntegrationTests/LoginTest.cs]: this comment still stands

File: Phaethon/Tests/IntegrationTests/LoginTest.cs [hunk 27]
```diff
@@ -165,15 +145,14 @@ public async Task Edit_User_OK()
...
+            _name2 = _userModel2.Username;
             //Act
-            var response = await _client.PostAsJsonAsync("Login/CreateOrUpdate", userModel2);
+            var response = await _client.PostAsJsonAsync("Login/CreateOrUpdate", _userModel2);
             //Assert
             Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
-            Assert.AreNotEqual(name, name2);
-            Cleanup();
+            Assert.AreNotEqual(_name, _name2);
         }
```
Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: I can't login in the webclient

Comment (RaidenRabit) [reply Phaethon/Tests/IntegrationTests/LoginTest.cs]: It keeps redirecting me to the Login Page

Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: Missing UI elements to show me that I'm logged in or not

Comment (RaidenRabit) [inline Phaethon/Tests/IntegrationTests/LoginTest.cs]: Your UI pages should be using the LanguagePack, as oposed to hardcoding in text. Have a look at any other cshtml file for reference on how to implement that, or just ask me/ralfs.

File: Phaethon/Core/Model/Login.cs (hunk 1 above)
Comment (bubriks) [reply Phaethon/Core/Model/Login.cs]: he could but thats alot of changes to hes current code and its way easyer to have fk to int not name (in total i kinda agree but i beleive its a bit too late to bring this issue up right now)

File: Phaethon/WebClient/Controllers/LoginController.cs [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+        public async Task<ActionResult> Index(Login loginModel)
+        {
+            var response = await _client.PostAsJsonAsync("Login", loginModel);
+            if (HttpStatusCode.OK == response.StatusCode)
+            {
+                var deserializedResponse = JsonConvert.DeserializeObject<int>(await response.Content.ReadAsStringAsync());
+                if (deserializedResponse != 0)
+                {
+                    Session["ID"] = deserializedResponse;
+                    Session["Username"] = loginModel.Username + " ";
```
Comment (bubriks) [inline Phaethon/WebClient/Controllers/LoginController.cs]: What is the pourpuse of Session "Username"?

File: Phaethon/WebClient/Views/Shared/_Layout.cshtml [hunk 29]
```diff
@@ -1,4 +1,19 @@
 ﻿@using LanguagePack = WebClient.Resources.Language_Files.LanguagePack;
+@{
+
+    var viewControlerName = ViewContext.RouteData.Values["controller"].ToString();
+    if (viewControlerName != "Login")
+    {
+        if (Session["ID"] == null)
```
Comment (bubriks) [inline Phaethon/WebClient/Views/Shared/_Layout.cshtml]: both of thease if stmts could be added to same if, so less space is taken and more readable, just add && between them :)

File: Phaethon/WebClient/Controllers/LoginController.cs (hunk 28 above)
Comment (AdamBlazsek) [reply Phaethon/WebClient/Controllers/LoginController.cs]: Session "Username" gives you the currently logged in person which is displayed on the top header
